wiring/src/spark_wiring_ble.cpp

```diff
@@ -577,81 +576,81 @@ BleAdvertisingData::BleAdvertisingData(const iBeacon& beacon)
     }
 
-    selfData_[selfLen_++] = 0x02;
-    selfData_[selfLen_++] = BLE_SIG_AD_TYPE_FLAGS;
-    selfData_[selfLen_++] = BLE_SIG_ADV_FLAGS_LE_ONLY_GENERAL_DISC_MODE;
-    selfData_[selfLen_++] = 0x1a; // length
-    selfData_[selfLen_++] = BLE_SIG_AD_TYPE_MANUFACTURER_SPECIFIC_DATA;
+    CHECK_TRUE(selfData_.reserve(BLE_MAX_ADV_DATA_LEN), 0);
+    selfData_.append(0x02);
+    selfData_.append(BLE_SIG_AD_TYPE_FLAGS);
+    selfData_.append(BLE_SIG_ADV_FLAGS_LE_ONLY_GENERAL_DISC_MODE);
+    selfData_.append(0x1a); // length
+    selfData_.append(BLE_SIG_AD_TYPE_MANUFACTURER_SPECIFIC_DATA);
     // Company ID
     uint16_t companyId = iBeacon::APPLE_COMPANY_ID;
-    memcpy(&selfData_[selfLen_], (uint8_t*)&companyId, sizeof(uint16_t));
-    selfLen_ += sizeof(uint16_t);
+    selfData_.append((const uint8_t *)&companyId, 2);
     // Beacon type: iBeacon
-    selfData_[selfLen_++] = iBeacon::BEACON_TYPE_IBEACON;
+    selfData_.append(iBeacon::BEACON_TYPE_IBEACON);
     // Length of the following payload
-    selfData_[selfLen_++] = 0x15;
+    selfData_.append(0x15);
     // 128-bits Beacon UUID, MSB
     for (size_t i = 0, j = BLE_SIG_UUID_128BIT_LEN - 1; i < BLE_SIG_UUID_128BIT_LEN; i++, j--) {
-        selfData_[selfLen_ + i] = beacon.UUID().rawBytes()[j];
+        selfData_.append(beacon.UUID().rawBytes()[j]);
     }
-    selfLen_ += BLE_SIG_UUID_128BIT_LEN;
     // Major, MSB
-    selfData_[selfLen_++] = (uint8_t)((beacon.major() >> 8) & 0x00FF);
-    selfData_[selfLen_++] = (uint8_t)(beacon.major() & 0x00FF);
+    selfData_.append((uint8_t)((beacon.major() >> 8) & 0x00FF));
+    selfData_.append((uint8_t)(beacon.major() & 0x00FF));
     // Minor, MSB
-    selfData_[selfLen_++] = (uint8_t)((beacon.minor() >> 8) & 0x00FF);
-    selfData_[selfLen_++] = (uint8_t)(beacon.minor() & 0x00FF);
+    selfData_.append((uint8_t)((beacon.minor() >> 8) & 0x00FF));
+    selfData_.append((uint8_t)(beacon.minor() & 0x00FF));
     // Measure power
-    selfData_[selfLen_++] = beacon.measurePower();
+    selfData_.append(beacon.measurePower());
+    selfData_.trimToSize();
 
-    return selfLen_;
+    return selfData_.size();
 }
 
 size_t BleAdvertisingData::append(BleAdvertisingDataType type, const uint8_t* buf, size_t len, bool force) {
     if (buf == nullptr) {
-        return selfLen_;
+        return selfData_.size();
     }
     size_t offset;
-    size_t adsLen = locate(selfData_, selfLen_, type, &offset);
+    size_t adsLen = locate(selfData_.data(), selfData_.size(), type, &offset);
     if (!force && adsLen > 0) {
         // Update the existing AD structure.
-        size_t staLen = selfLen_ - adsLen;
-        if ((staLen + len + 2) <= BLE_MAX_ADV_DATA_LEN) {
-            // Firstly, move the last consistent block.
-            uint16_t moveLen = selfLen_ - offset - adsLen;
-            memmove(&selfData_[offset + len + 2], &selfData_[offset + adsLen], moveLen);
+        if ((selfData_.size() - adsLen + len + 2) <= BLE_MAX_ADV_DATA_LEN_EXT) {
```

Comment:
We should use `BLE_MAX_ADV_DATA_LEN_EXT_CONNECTABLE` here.

Reply:
Application could be non-connectable.